### PR TITLE
Draft of docs for slashing

### PR DESCRIPTION
**Pull request by szotov**

### CODE REVIEW

File: docs/source/architecture/contracts.md [hunk 1]
```diff
@@ -15,14 +16,15 @@
 1. Deploy `NuCypherToken` with all future supply tokens
 2. Deploy `MinersEscrow` with a dispatcher targeting it
 3. Deploy `PolicyManager` with its own dispatcher, also targeting it
-4. Transfer reward tokens to the `MinersEscrow` contract. These tokens are future mining rewards, and initial allocations
-5. Run the `initialize()` method to initialize the `MinersEscrow` contract
-6. Set the address of the `PolicyManager` contract  in the `MinersEscrow` by using the `setPolicyManager(address)`
-7. Pre-deposit tokens to the `MinersEscrow` if necessary:
+4. Deploy `MiningAdjudicator` with a dispatcher
+5. Transfer reward tokens to the `MinersEscrow` contract. These tokens are future mining rewards, and initial allocations
```
Comment (mswilkison) [inline docs/source/architecture/contracts.md]:
```suggestion
5. Transfer reward tokens to the `MinersEscrow` contract. These tokens are future mining rewards and initial allocations
```

File: docs/source/architecture/contracts.md [hunk 2]
```diff
@@ -15,14 +16,15 @@
...
+6. Run the `initialize()` method to initialize the `MinersEscrow` contract
+7. Set the address of the `PolicyManager` contract  in the `MinersEscrow` by using the `setPolicyManager(address)`
+8. Pre-deposit tokens to the `MinersEscrow` if necessary:
 	* Approve the transfer tokens for the `MinersEscrow` contract using the `approve(address, uint)` method. The parameters are the address of `MinersEscrow` and the amount of tokens for a miner or group of miners;
 	* Deposit tokens to the `MinersEscrow` contract using the `preDeposit(address[], uint[], uint[])` method. The parameters are the addresses of the miners, the amount of tokens for each miner and the periods during which tokens will be locked for each miner
```
Comment (mswilkison) [inline docs/source/architecture/contracts.md]:
```suggestion
	* Deposit tokens to the `MinersEscrow` contract using the `preDeposit(address[], uint[], uint[])` method. The parameters are the addresses of the miners, the amount of tokens for each miner, and the periods during which tokens will be locked for each miner
```

File: docs/source/architecture/contracts.md [hunk 3]
```diff
@@ -15,14 +16,15 @@
...
-8. Deploy `UserEscrowProxy` with `UserEscrowLibraryLinker` targeting it
-9. Pre-deposit tokens to the `UserEscrow`, and if necessary:
+9. Deploy `UserEscrowProxy` with `UserEscrowLibraryLinker` targeting it
+10. Pre-deposit tokens to the `UserEscrow`, and if necessary:
```
Comment (mswilkison) [inline docs/source/architecture/contracts.md]:
```suggestion
10. Pre-deposit tokens to the `UserEscrow` and, if necessary:
```

File: docs/source/architecture/contracts.md [hunk 4]
```diff
@@ -15,14 +16,15 @@
...
```
Comment (mswilkison) [inline docs/source/architecture/contracts.md]:
```suggestion
```

File: docs/source/architecture/contracts.md [hunk 5]
```diff
@@ -15,14 +16,15 @@
...
 
 * Create new instance of the `UserEscrow` contract
```
Comment (mswilkison) [inline docs/source/architecture/contracts.md]:
```suggestion
	* Create new instance of the `UserEscrow` contract 
```

File: docs/source/architecture/contracts.md [hunk 6]
```diff
@@ -15,14 +16,15 @@
...
+8. Pre-deposit tokens to the `MinersEscrow` if necessary:
 	* Approve the transfer tokens for the `MinersEscrow` contract using the `approve(address, uint)` method. The parameters are the address of `MinersEscrow` and the amount of tokens for a miner or group of miners;
 	* Deposit tokens to the `MinersEscrow` contract using the `preDeposit(address[], uint[], uint[])` method. The parameters are the addresses of the miners, the amount of tokens for each miner and the periods during which tokens will be locked for each miner
-8. Deploy `UserEscrowProxy` with `UserEscrowLibraryLinker` targeting it
-9. Pre-deposit tokens to the `UserEscrow`, and if necessary:
+9. Deploy `UserEscrowProxy` with `UserEscrowLibraryLinker` targeting it
+10. Pre-deposit tokens to the `UserEscrow`, and if necessary:
 
 * Create new instance of the `UserEscrow` contract 
 * Transfer ownership of the instance of the `UserEscrow` contract to the user
```
Comment (mswilkison) [inline docs/source/architecture/contracts.md]:
```suggestion
	* Transfer ownership of the instance of the `UserEscrow` contract to the user
```

File: docs/source/architecture/slashing.rst [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+--------------------------
+
+The punishment is to reduce the number of tokens that belongs to the offending staker.
+In this case, the main task is not to violate the logic of locking tokens.
+The whole stake consists of several parts:
+
+* tokens which the staker can withdraw at any moment
+* tokens locked for a specific period
+
+Since a staker can divide a stake (extend a part of a stake), as well as lock new tokens, then the total stake is represented as the sum of all the sub stakes active at a particular period. Each sub stake has a beginning and duration, which is reduced only upon confirmation of the activity. Staker slashed on the principle: first a sub stake, which will end earlier. Therefore, in the first place, the not-locked part of tokens belonging to the staker is slashed. After that, if necessary, sub stakes are adjusted: finds the shortest sub stake and decreases by the required amount. If this is not enough, then the next short sub stake is searched for, and so on. A sub stake that begins in the next period is checked separately.
```
Comment (cygnusv) [inline docs/source/architecture/slashing.rst]:
```suggestion
Since a staker can divide a stake (extend a part of a stake), as well as lock new tokens, then the total stake is represented as the sum of all the sub stakes active at a particular period. Each sub stake has a beginning and duration, which is reduced only upon confirmation of the activity. Staker slashed on the principle: first a sub stake, which will end earlier. Therefore, in the first place, the unlocked part of tokens belonging to the staker is slashed. After that, if necessary, sub stakes are adjusted: the shortest sub stake is found and decreased by the required amount; if this is not enough, then the next short sub stake is searched for, and so on. A sub stake that begins in the next period is checked separately.
```

File: docs/source/architecture/slashing.rst [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+Example:
+
+1000 tokens belong to the staker, 1st sub stake - 500 tokens are locked for 10 periods, 2nd sub stake - 200 tokens for 2 periods and 3rd sub stake - 100 tokens that are locked starting from the next period.
+
+* Penalty 100 tokens. The parameters of sub stakes will not change. Only the number of tokens belonging to the staker will decrease from 1000 to 900. 
+* Penalty 400 tokens. There will be 600 tokens belonging to the staker. At the same time, in the current period 700 tokens were locked, and in the next only 600. So to normalize stake distribution - the shortest sub stake (second) is reduced by 100 tokens, the remaining sub stakes remain unchanged
+* Penalty 600 tokens. 400 tokens are remain. Reducing the shortest sub stake (second) is not enough, so it's removed. After the first sub stake is reduced from 500 to 400, and the third sub stake which starting in the next period is also removed.
```
Comment (cygnusv) [inline docs/source/architecture/slashing.rst]:
```suggestion
* Penalty 600 tokens. 400 tokens remain. Reducing the shortest sub stake (second) is not enough, so it's removed. After the first sub stake is reduced from 500 to 400, and the third sub stake which starting in the next period is also removed.
```
In any case, I don't understand the last sentence ("`After the first sub stake is reduced from 500 to 400, and the third sub stake which starting in the next period is also removed.`"). Can you rephrase it? It feels like it's incomplete.

Comment (szotov) [reply docs/source/architecture/slashing.rst]: How about this:
`After removing the second sub stake we have this locking: 500 tokens in the current period and 600 tokens in the next one. Next step is reducing the first sub stake from 500 to 400. Now current period has correct locking (400 tokens) but the next period still has locked tokens more than staker has (500). To adjust this the third sub stake must be removed.`

File: docs/source/architecture/slashing.rst [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+The whole stake consists of several parts:
+
+* tokens which the staker can withdraw at any moment
+* tokens locked for a specific period
+
+Since a staker can divide a stake (extend a part of a stake), as well as lock new tokens, then the total stake is represented as the sum of all the sub stakes active at a particular period. Each sub stake has a beginning and duration, which is reduced only upon confirmation of the activity. Staker slashed on the principle: first a sub stake, which will end earlier. Therefore, in the first place, the unlocked part of tokens belonging to the staker is slashed. After that, if necessary, sub stakes are adjusted: the shortest sub stake is found and decreased by the required amount; if this is not enough, then the next short sub stake is searched for, and so on. A sub stake that begins in the next period is checked separately.
+
+Example:
+
+1000 tokens belong to the staker, 1st sub stake - 500 tokens are locked for 10 periods, 2nd sub stake - 200 tokens for 2 periods and 3rd sub stake - 100 tokens that are locked starting from the next period.
```
Comment (derekpierre) [inline docs/source/architecture/slashing.rst]:
```suggestion
A staker has 1000 tokens:

* 1st sub stake = 500 tokens locked for 10 periods
* 2nd sub stake = 200 tokens for 2 periods
* 3rd sub stake = 100 tokens locked starting from the next period
* Remaining 200 tokens
```

File: docs/source/architecture/slashing.rst [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+* Penalty 100 tokens. The parameters of sub stakes will not change. Only the number of tokens belonging to the staker will decrease from 1000 to 900.
```
Comment (derekpierre) [inline docs/source/architecture/slashing.rst]:
```suggestion
Penalty Scenarios:

* *100 tokens*:
 
 The values of sub stakes will not change and unlocked tokens can be used. Only the remaining tokens will be reduced from 200 to 100.

 Result:
   * 1st sub stake = 500 tokens locked for 10 periods
   * 2nd sub stake = 200 tokens for 2 periods
   * 3rd sub stake = 100 tokens locked starting from the next period
   * Remaining 100 tokens
```

File: docs/source/architecture/slashing.rst [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+Example:
+
+1000 tokens belong to the staker, 1st sub stake - 500 tokens are locked for 10 periods, 2nd sub stake - 200 tokens for 2 periods and 3rd sub stake - 100 tokens that are locked starting from the next period.
+
+* Penalty 100 tokens. The parameters of sub stakes will not change. Only the number of tokens belonging to the staker will decrease from 1000 to 900. 
+* Penalty 400 tokens. There will be 600 tokens belonging to the staker. At the same time, in the current period 700 tokens were locked, and in the next only 600. So to normalize stake distribution - the shortest sub stake (second) is reduced by 100 tokens, the remaining sub stakes remain unchanged
```
Comment (derekpierre) [inline docs/source/architecture/slashing.rst]:
```suggestion

* *400 tokens*:

 The remaining tokens can only cover 200 tokens. In the current period, 700 tokens were locked but the 3rd sub stake has not yet been locked. The 3rd sub stake can only cover 100 tokens which is still not enough. Therefore, the shortest locked sub stake (2nd) is reduced by 100 tokens and the other sub stakes remain unchanged.

 Result:
   * 1st sub stake = 500 tokens locked for 10 periods
   * 2nd sub stake = 100 tokens for 2 periods
   * 3rd sub stake = 0 tokens locked starting from the next period
   * Remaining 0 tokens
```

File: docs/source/architecture/slashing.rst [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+* Penalty 600 tokens. 400 tokens remain. Reducing the shortest sub stake (second) is not enough, so it's removed. After the first sub stake is reduced from 500 to 400, and the third sub stake which starting in the next period is also removed.
```
Comment (derekpierre) [inline docs/source/architecture/slashing.rst]:
```suggestion
 
* *600 tokens*:

  Reducing the unlocked remaining tokens, 3rd sub stakes, and the shortest sub stake (2nd) is not enough, so they are all removed. The next shortest sub stake is the 1st which is reduced from 500 to 400. 

 Result:
   * 1st sub stake = 400 tokens locked for 10 periods
   * 2nd sub stake = 0 tokens for 2 periods
   * 3rd sub stake = 0 tokens locked starting from the next period
   * Remaining 0 tokens
```

File: docs/source/architecture/contracts.md [hunk 13]
```diff
@@ -6,6 +6,7 @@
 * `NuCypherToken` ERC20 token contract
 * `MinersEscrow` Holds Ursula's stake, stores information about Ursula's activity, and assigns a reward for participating in the NuCypher network. (The `Issuer` contract is part of the `MinersEscrow`)
 * `PolicyManager` Holds a policy's fee and distributes fee by periods
+* `MiningAdjudicator` Manages [the slashing protocol](slashing)
 * `Upgradeable` Base contract for [upgrading](upgradeable_proxy_contracts)
 * `Dispatcher` Proxy to other contracts. This provides upgrading of the `MinersEscrow` and `PolicyManager` contracts
```
[the remark this answers is not in the document]
Comment (szotov) [reply docs/source/architecture/contracts.md]: "that" refers to "proxy" or "other contracts"?

[the remark this answers is not in the document]
Comment (cygnusv) [reply docs/source/architecture/contracts.md]: It refers to "Proxy", as far as I understand. Is that correct?

File: docs/source/architecture/contracts.md [hunk 14]
```diff
@@ -15,19 +16,19 @@
...
 3. Deploy `PolicyManager` with its own dispatcher, also targeting it
-4. Transfer reward tokens to the `MinersEscrow` contract. These tokens are future mining rewards, and initial allocations
-5. Run the `initialize()` method to initialize the `MinersEscrow` contract
-6. Set the address of the `PolicyManager` contract  in the `MinersEscrow` by using the `setPolicyManager(address)`
-7. Pre-deposit tokens to the `MinersEscrow` if necessary:
+4. Deploy `MiningAdjudicator` with a dispatcher
+5. Transfer reward tokens to the `MinersEscrow` contract. These tokens are future mining rewards and initial allocations
+6. Run the `initialize()` method to initialize the `MinersEscrow` contract
+7. Set the address of the `PolicyManager` contract  in the `MinersEscrow` by using the `setPolicyManager(address)`
+8. Pre-deposit tokens to the `MinersEscrow` if necessary:
```
[the remark this answers is not in the document]
Comment (szotov) [reply docs/source/architecture/contracts.md]: Initial thought was these additional step is not necessary for correct work. But if needed then do this steps. So how to better formulate?

File: docs/source/architecture/slashing.rst [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+The punishment is to reduce the number of tokens that belongs to the offending staker.
+In this case, the main task is not to violate the logic of locking tokens.
+The whole stake consists of several parts:
+
+* tokens which the staker can withdraw at any moment
+* tokens locked for a specific period
+
+Since a staker can divide a stake (extend a part of a stake), as well as lock new tokens, then the total stake is represented as the sum of all the sub stakes active at a particular period. Each sub stake has a beginning and duration, which is reduced only upon confirmation of the activity. Staker slashed on the principle: first a sub stake, which will end earlier. Therefore, in the first place, the unlocked part of tokens belonging to the staker is slashed. After that, if necessary, sub stakes are adjusted: the shortest sub stake is found and decreased by the required amount; if this is not enough, then the next short sub stake is searched for, and so on. A sub stake that begins in the next period is checked separately.
+
+Example:
```
[the remark this answers is not in the document]
Comment (szotov) [reply docs/source/architecture/slashing.rst]: You are almost right, I just changed some phrases about 3rd sub stake. Thanks for big review

File: docs/source/architecture/contracts.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (derekpierre) [reply docs/source/architecture/contracts.md]: Yes it refers to `Proxy`, but a fair point that it is a bit unclear based on the wording. Perhaps `and` can be used instead...?
```suggestion
* `Dispatcher` Proxy to other contracts and provides upgrading of the `MinersEscrow` and `PolicyManager` contracts
```

File: docs/source/architecture/slashing.rst (hunk 8 above)
Comment (szotov) [reply docs/source/architecture/slashing.rst]: I think, in current changes it's more clear, look at

File: docs/source/architecture/slashing.rst [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+--------------------------
+
+The punishment is to reduce the number of tokens that belongs to the offending staker.
+In this case, the main task is not to violate the logic of locking tokens.
+The entire stake consists of:
+
+* tokens which the staker can withdraw at any moment
+* tokens locked for a specific period
+
+Since a staker can divide a stake (extend a part of a stake), as well as lock new tokens, then the total stake is represented as the sum of all the sub stakes active at a particular period. Each sub stake has a beginning and duration, which is reduced only upon confirmation of the activity. Sub stakes get slashed in order based on their duration, beginning with the shortest. Therefore, to start, the unlocked part of tokens belonging to the staker is slashed. After that, if necessary, sub stakes are decreased based on duration: the shortest sub stake is decreased by the required amount; if the adjustment of the sub stake is not enough, then the next shortest sub stake is decreased, and so on. A sub stake that begins in the next period is checked separately.
```
Comment (KPrasch) [inline docs/source/architecture/slashing.rst]: I have grown confused regarding the "divide stake" language.

Here is says:

> a staker can divide a stake (extend a part of a stake)

Which exemplifies my confusion.  Dive to my ear sounds like a reduction, the opposite of expansion - So how can "dive stake" indicate "extend stake"?

File: docs/source/architecture/slashing.rst [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+**Example:**
```
Comment (KPrasch) [inline docs/source/architecture/slashing.rst]: I'd love to see some ascii art in here :-)

File: docs/source/architecture/slashing.rst (hunk 16 above)
Comment (KPrasch) [reply docs/source/architecture/slashing.rst]: How can Ursula discover that her stake has been slashed?

File: docs/source/architecture/slashing.rst [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+--------------------------------
+
+TBD (https://github.com/nucypher/nucypher/issues/803)
+
+
+How slashing affects stake
+--------------------------
+
+The punishment is to reduce the number of tokens that belongs to the offending staker.
+In this case, the main task is not to violate the logic of locking tokens.
```
[the remark this answers is not in the document]
Comment (szotov) [reply docs/source/architecture/slashing.rst]: Another meaning - when we slash Ursula we also should not break logic of sub stakes. If we will not decrease amount of sub stake then Ursula will get reward more than she deserve.

File: docs/source/architecture/slashing.rst [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+The entire stake consists of:
```
[the remark this answers is not in the document]
Comment (szotov) [reply docs/source/architecture/slashing.rst]: I want to pay attention that one Ursula has one stake (and multiple sub stakes). So changing `entire` to `one` still has the same meaning?

File: docs/source/architecture/slashing.rst (hunk 16 above)
Comment (szotov) [reply docs/source/architecture/slashing.rst]: I meant divide stake (or split stake). How to rephrase that?
Ursula can check `Slashed` event

File: docs/source/architecture/contracts.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (derekpierre) [reply docs/source/architecture/contracts.md]: `if necessary` is probably fine then.

File: docs/source/architecture/slashing.rst [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+
+A staker has 1000 tokens:
+	* 1st sub stake = 500 tokens locked for 10 periods
+	* 2nd sub stake = 200 tokens for 2 periods
+	* 3rd sub stake = 100 tokens locked starting from the next period
+	* Remaining 200 tokens
+
+Penalty Scenarios:
+
+* *100 tokens*:
```
Comment (derekpierre) [inline docs/source/architecture/slashing.rst]: Perhaps we can do something here for the scenario to stand out a bit better.

Maybe:
```suggestion
* *SCENARIO 1: 100 tokens*:
```
or something like that, and similar for the others (SCENARIO 2: 400 tokens, etc.). The bullet points for each scenario don't stand out enough in my mind. Thoughts?

File: docs/source/architecture/slashing.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (arjunhassard) [reply docs/source/architecture/slashing.rst]:
```suggestion
However, in doing so, the protocol must not violate the logic governing the lock periods associated with a stake. 
```

File: docs/source/architecture/slashing.rst (hunk 19 above)
[the remark this answers is not in the document]
Comment (arjunhassard) [reply docs/source/architecture/slashing.rst]: I'd go with 'entire' to convey that it's a composite

File: docs/source/architecture/slashing.rst (hunk 20 above)
Comment (arjunhassard) [reply docs/source/architecture/slashing.rst]: @derekpierre Definitely. Going further, I'd remind the reader that a punishment is due. 
`Scenario 1: Staker incurs penalty calculated to be worth *100 tokens*:`

File: docs/source/architecture/slashing.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (szotov) [reply docs/source/architecture/slashing.rst]: I hear emphasis on duration, but only amount of sub stakes are adjusted. Maybe I'm wrong

File: docs/source/architecture/slashing.rst [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+	Result:
+
+		* 1st sub stake = 500 tokens locked for 10 periods
+		* 2nd sub stake = 200 tokens for 2 periods
+		* 3rd sub stake = 100 tokens locked starting from the next period
+		* Remaining 100 tokens
+   
+* *Scenario 2*: Staker incurs penalty calculated to be worth **400 tokens**:
+
+	The remaining tokens can only cover 200 tokens. In the current period, 700 tokens were locked and 800 tokens were locked in the next one. The 3rd sub stake can only cover 100 tokens for the lock in  next period which is still not enough. Therefore, the shortest locked sub stake (2nd) is reduced by 100 tokens and the other sub stakes remain unchanged.
```
Comment (derekpierre) [inline docs/source/architecture/slashing.rst]: Could we update the wording here a little...?
```suggestion
	The remaining tokens can only cover 200 tokens. In the current period, 700 tokens were locked and 800 tokens were locked in the next one. The 3rd sub stake is not yet locked (not until the next period) and is available for slashing but can only cover 100 tokens which are still not enough. Therefore, the shortest locked sub stake (2nd) is reduced by 100 tokens and the other sub stakes remain unchanged.
```

Comment (szotov) [reply docs/source/architecture/slashing.rst]: It's a little bit wrong. I want to explain "locked in the next period": when Ursula call `lock` - a sub stake is created with `firstPeriod` == `nextPeriod`, because if use current period then Ursula will get reward also for this lock in the current period and this is unfair (she just locked tokens). So lock will start "work" for Ursula in the next period but even in the current period Ursula can't withdraw them and techically they are already locked.
I see that is blind spot in explanation. Help me, where better to add something to help with understanding?

Comment (derekpierre) [reply docs/source/architecture/slashing.rst]: Ah ok. So technically the 3rd sub stake has been locked just not for this period, but since it's work hasn't been started as yet, it is considered for slashing before other sub stakes already locked for work...

```suggestion
	The remaining tokens can only cover 200 tokens. In the current period, 700 tokens are locked and 800 tokens are locked for the next period. The 3rd sub stake is locked for the next period but has not yet been used as a deposit for "work" - not until the next period begins. Therefore, it will be the next sub stake to be slashed. However, it can only cover 100 tokens for the penalty which is still not enough. The next shortest locked sub stake (2nd) is then reduced by 100 tokens to cover the remainder of the penalty, and the other sub stakes remain unchanged.
```

File: docs/source/architecture/slashing.rst [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+	Result:
+
+		* 1st sub stake = 500 tokens locked for 10 periods
+		* 2nd sub stake = 200 tokens for 2 periods
+		* 3rd sub stake = 100 tokens locked starting from the next period
+		* Remaining 100 tokens
+   
+* *Scenario 2*: Staker incurs penalty calculated to be worth **400 tokens**:
+
+	The remaining tokens can only cover 200 tokens. In the current period, 700 tokens are locked and 800 tokens are locked for the next period. The 3rd sub stake is locked for the next period but has not yet been used as a deposit for "work" - not until the next period begins. Therefore, it will be the next sub stake to be slashed. However, it can only cover 100 tokens for the penalty which is still not enough. The next shortest locked sub stake (2nd) is then reduced by 100 tokens to cover the remainder of the penalty, and the other sub stakes remain unchanged.
```
Comment (KPrasch) [inline docs/source/architecture/slashing.rst]: I find this line a bit confusing: "The remaining tokens can only cover 200 tokens."

File: docs/source/architecture/slashing.rst [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+	* 1st sub stake = 500 tokens locked for 10 periods
+	* 2nd sub stake = 200 tokens for 2 periods
+	* 3rd sub stake = 100 tokens locked starting from the next period
+	* 200 tokens in an unlocked state (still staked, but can be freely withdrawn).
+
+Penalty Scenarios:
+
+* *Scenario 1*: Staker incurs penalty calculated to be worth **100 tokens**:
+
+	The values of sub stakes will not change and unlocked tokens can be used. Only the remaining tokens will be reduced from 200 to 100.
```
Comment (arjunhassard) [inline docs/source/architecture/slashing.rst]:
```suggestion
	Only the unlocked tokens will be reduced; from 200 to 100. The values of locked sub-stakes will therefore remain unchanged in this punishment scenario. 
```

File: docs/source/architecture/slashing.rst [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+
+	Result:
+
+		* 1st sub stake = 500 tokens locked for 10 periods
+		* 2nd sub stake = 200 tokens for 2 periods
+		* 3rd sub stake = 100 tokens locked starting from the next period
+		* Remaining 100 tokens
```
Comment (arjunhassard) [inline docs/source/architecture/slashing.rst]:
```suggestion
		* 100 tokens in an unlocked state
```

File: docs/source/architecture/slashing.rst (hunk 22 above)
Comment (arjunhassard) [reply docs/source/architecture/slashing.rst]: It's confusing because the word 'remaining' is used to describe both freely withdrawable tokens, and tokens that have been locked within the current cycle, but are not active until the next day. I think there are too many fiddly concepts layered on top of each other within one Scenario here. I suggest adding another scenario, and order all the scenarios such that they provide examples of successively less important ideas (i.e. the fact that tokens don't become active for up to 24h is less important than the fact that shorter sub-stakes are slashed first)

Comment (szotov) [reply docs/source/architecture/slashing.rst]: I tried to explain slashing logic on high level but have made mistake in scenario 2 and confused you all.
So I added new scenario and described a little more deeply + some diagrams

File: docs/source/architecture/slashing.rst (hunk 17 above)
Comment (szotov) [reply docs/source/architecture/slashing.rst]: 👍

Comment (KPrasch) [reply docs/source/architecture/slashing.rst]: YES.

File: docs/source/architecture/contracts.md [hunk 25]
```diff
@@ -6,28 +6,29 @@
...
 	* Approve the transfer tokens for the `MinersEscrow` contract using the `approve(address, uint)` method. The parameters are the address of `MinersEscrow` and the amount of tokens for a miner or group of miners;
-	* Deposit tokens to the `MinersEscrow` contract using the `preDeposit(address[], uint[], uint[])` method. The parameters are the addresses of the miners, the amount of tokens for each miner and the periods during which tokens will be locked for each miner
-8. Deploy `UserEscrowProxy` with `UserEscrowLibraryLinker` targeting it
-9. Pre-deposit tokens to the `UserEscrow`, and if necessary:
-
-* Create new instance of the `UserEscrow` contract 
-* Transfer ownership of the instance of the `UserEscrow` contract to the user
-* Approve the transfer of tokens for the `UserEscrow`
-* Deposit tokens by the `initialDeposit(uint256, uint256)` method
+	* Deposit tokens to the `MinersEscrow` contract using the `preDeposit(address[], uint[], uint[])` method. The parameters are the addresses of the miners, the amount of tokens for each miner, and the periods during which tokens will be locked for each miner
```
Comment (michwill) [inline docs/source/architecture/contracts.md]: the _number of periods_?
(Otherwise, could be understood as: all periods are numbered, and you specify period numbers when it's locked, which is a bit confusing)

File: docs/source/architecture/slashing.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (michwill) [reply docs/source/architecture/slashing.rst]: I think, I didn't really understand what this sentence is trying to say. Neither in @szotov, nor in @arjunhassard variant

File: docs/source/architecture/slashing.rst (hunk 19 above)
[the remark this answers is not in the document]
Comment (michwill) [reply docs/source/architecture/slashing.rst]: +1 for the entire

File: docs/source/architecture/slashing.rst [hunk 26]
```diff
@@ -0,0 +1,150 @@
...
+    * tokens locked for a specific period
+
+A staker may extend the unlock period for any number of portions of their total stake. This divides the stake into smaller parts, each with a unique unlock date in the future. Stakers may also acquire and lock new tokens. The total stake is represented as the sum of all the different sub-stakes active in a given cycle (new cycle every 24h), which includes locked sub-stakes, and any sub-stakes that have passed their unlock date, and can be freely withdrawn. Each sub-stake has a beginning and duration (lock time). When a staker confirms activity each day, the remaining lock time for relevant sub-stakes is reduced. 
+
+Sub stakes get slashed in the order of their remaining lock time, beginning with the shortest – so the first portion of the stake to be slashed is the unlocked portion. After that, if necessary, locked sub-stakes are decreased – the shortest sub stake is decreased by the required amount; if the adjustment of that sub-stake is insufficient to fulfil the required punishment sum, then the next shortest sub-stake is decreased, and so on. Sub-stakes that begin in the next period are checked separately.
```
Comment (michwill) [inline docs/source/architecture/slashing.rst]: I like the language of this and previous paragraph. Very clear and understandable!

File: docs/source/architecture/slashing.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (szotov) [reply docs/source/architecture/slashing.rst]: This part of explanation is important, so need to rephrase before merge. The idea is from slashing perspective sub stakes are not hold information about how many tokens belongs to a stake. Sub stakes define when staker will get tokens back but also how big will be reward. Bigger sub stakes does not mean that staker has more tokens. So if we decrease amount of belonging tokens from 1000 to 900 and sum of sub stakes is still 1000 - then miner will get more reward for some periods. As a result decreasing sum of sub stakes to amount of tokens will be sufficient. This what I meant - do not breake logic of locking tokens

File: docs/source/architecture/contracts.md (hunk 25 above)
Comment (szotov) [reply docs/source/architecture/contracts.md]: 👍

File: docs/source/architecture/slashing.rst (hunk 26 above)
Comment (KPrasch) [reply docs/source/architecture/slashing.rst]: A bit of a run-on sentence here: 

"The total stake is represented as the sum of all the different sub-stakes active in a given cycle (new cycle every 24h), which includes locked sub-stakes, and any sub-stakes that have passed their unlock date, and can be freely withdrawn."

How about:

"The total stake value is the sum of all active sub-stake values in a given period (new period every 24h).  This includes locked sub-stakes and any sub-stakes that have passed their unlock date which can be freely withdrawn."

File: docs/source/architecture/slashing.rst [hunk 27]
```diff
@@ -0,0 +1,150 @@
...
+
+**Example:**
+
+    A staker has 1000 tokens:
+        * 1st sub stake = 500 tokens locked for 10 periods
+        * 2nd sub stake = 200 tokens for 2 periods
+        * 3rd sub stake = 100 tokens locked starting from the next period and locked for 5 periods. The 3rd sub stake is locked for the next period but has not yet been used as a deposit for "work" - not until the next period begins.
+        * 200 tokens in an unlocked state (still staked, but can be freely withdrawn).
+
+    .. aafig::
```
Comment (KPrasch) [inline docs/source/architecture/slashing.rst]: :boom:  - I love the visuals!

File: docs/source/architecture/slashing.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (michwill) [reply docs/source/architecture/slashing.rst]: Ok, this is how I understand (for others to have the same understanding too):
"Only part of the tokens held in the mining contract are locked in sub-stakes (others are available to withdraw: they could be from uncollected staking compensation). When slashing is happening, we first decrease the total amount of tokens held by the contract, and then adjust the sub-stakes so that they hold no less tokens than Ursula has in the contract. So, if misbehavior occurs, Ursula firstly loses her liquid (non-withdrawn) tokens, and then the tokens locked for staking"

[the remark this answers is not in the document]
Comment (mswilkison) [reply docs/source/architecture/slashing.rst]: @michwill if your interpretation is correct it seems counterintuitive that you can lose non-staked tokens...

[the remark this answers is not in the document]
Comment (michwill) [reply docs/source/architecture/slashing.rst]: That's my understanding based on conversation with @szotov. And that's why I translated it here for discussion: it may indeed seem counter-intuitive!
So what do we want? Lose only staked tokens, or anything you may have sitting in that smart contract?

[the remark this answers is not in the document]
Comment (szotov) [reply docs/source/architecture/slashing.rst]: Non-staked tokens is unlocked sub stakes plus reward.
Of course if it's wrong way and should slash only locked tokens - I fix this in code and docs

File: docs/source/architecture/slashing.rst (hunk 26 above)
Comment (szotov) [reply docs/source/architecture/slashing.rst]: I realized that meaning of "stake" is a little bit different in parts of this doc. Is it all tokens that staker put as a stake in the contract and doesn't matter them are locked or not in the current period? Or only locked tokens are "stake"?

Comment (KPrasch) [reply docs/source/architecture/slashing.rst]: Hmm - I think.. only locked tokens. Perhaps we need a term for the tokens in escrow that are not locked?

Comment (arjunhassard) [reply docs/source/architecture/slashing.rst]: @KPrasch In that case, it seems a little odd that unlocked tokens sitting around can be slashed for misdeeds, but do not count at all towards your total stake (and commensurate payouts). If you stop earning fees/rewards on tokens once they reach their unlock date, then there is zero upside to leaving them in that state. If we want to rush stakers into re-staking the tokens for a new lock period, or withdraw asap – which makes sense from a user point of view, as stakes with no lock period can unpredictably disappear, which may impede sharing policies – then I suppose this is the right approach. No matter what, these rules need to be extremely clear to stakers.

File: docs/source/conf.py [hunk 28]
```diff
@@ -51,6 +51,7 @@
     'sphinx.ext.napoleon',
     'sphinx.ext.mathjax',
     'sphinx.ext.viewcode',
+	'aafigure.sphinxext',
```
Comment (tuxxy) [inline docs/source/conf.py]:
```suggestion
    'aafigure.sphinxext',
```

File: docs/source/architecture/slashing.rst (hunk 18 above)
[the remark this answers is not in the document]
Comment (szotov) [reply docs/source/architecture/slashing.rst]: @michwill, @mswilkison I'm waiting your thoughts on that

[the remark this answers is not in the document]
Comment (mswilkison) [reply docs/source/architecture/slashing.rst]: I suppose if someone begins producing malicious re-encryptions and they have unclaimed rewards, there isn't anything wrong with slashing those rewards even though it's counter-intuitive. It just provides an incentive not leave rewards sitting there and to either withdraw them or restake.

[the remark this answers is not in the document]
Comment (michwill) [reply docs/source/architecture/slashing.rst]: Yeah, I think, I'd be ok with it, too
